examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1091 @@
+        smoother->initialize(mg_matrices, smoother_data);
+        smoother->set_steps(2);
+        return smoother;
+      }
+    else
+      AssertThrow(false, ExcNotImplemented());
+  }
+
+
+  template <int dim>
+  void AdvectionProblem<dim>::solve()
+  {
+    Timer time;
+
+    const double       solve_tol = 1e-8 * system_rhs.l2_norm();
+    const unsigned int max_iters = 200;
+    SolverControl      solver_control(max_iters, solve_tol, true, true);
+    solver_control.enable_history_data();
+
+    typedef MGTransferPrebuilt<Vector<double>> Transfer;
+    Transfer                                   mg_transfer(mg_constrained_dofs);
+    mg_transfer.build_matrices(dof_handler);
+
+    FullMatrix<double> coarse_matrix;
+    coarse_matrix.copy_from(mg_matrices[0]);
+    MGCoarseGridHouseholder<double, Vector<double>> coarse_grid_solver;
+    coarse_grid_solver.initialize(coarse_matrix);
+
+    std::unique_ptr<MGSmoother<Vector<double>>> mg_smoother = create_smoother();
+
+    mg::Matrix<Vector<double>> mg_matrix(mg_matrices);
+    mg::Matrix<Vector<double>> mg_interface_matrix_in(mg_interface_in);
+    mg::Matrix<Vector<double>> mg_interface_matrix_out(mg_interface_out);
+
+    Multigrid<Vector<double>> mg(
+      mg_matrix, coarse_grid_solver, mg_transfer, *mg_smoother, *mg_smoother);
+    mg.set_edge_matrices(mg_interface_matrix_out, mg_interface_matrix_in);
+
+    PreconditionMG<dim, Vector<double>, Transfer> preconditioner(dof_handler,
+                                                                 mg,
+                                                                 mg_transfer);
+
+
+    std::cout << "     Solving with GMRES to tol " << solve_tol << "..."
+              << std::endl;
+    SolverGMRES<> solver(solver_control);
+
+    time.restart();
+    solver.solve(system_matrix, solution, system_rhs, preconditioner);
+    time.stop();
+
+    std::cout << "          converged in " << solver_control.last_step()
+              << " iterations"
+              << " in " << time.last_wall_time() << " seconds " << std::endl;
+
+    constraints.distribute(solution);
+  }
+
+
+
+  template <int dim>
+  void AdvectionProblem<dim>::output_results(const unsigned int cycle) const
+  {
+    DataOut<dim> data_out;
+    data_out.attach_dof_handler(dof_handler);
+    data_out.add_data_vector(solution, "solution");
+
+    // Here we generate an index for each cell to visualize the ordering used
+    // by the smoothers. Note that we do this only for the active cells
+    // instead of the levels, where the smoothers are actually used. For the
+    // point smoothers we renumber DoFs instead of cells, so this is only an
+    // approximation of what happens in reality. Finally, the random ordering
+    // is not the random ordering we actually use (see create_smoother() for
+    // that).
+    const unsigned int n_active_cells = triangulation.n_active_cells();
+    Vector<double>     cell_indices(n_active_cells);
```

Comment:
you have an ordering issue here where the `cell_indices` variable goes out of scope before the `data_out` object declared above, but the latter has a reference to the former. Just move the declaration of the `DataOut` object and the two following lines (ll. 1001-1003) further down below, just before we actually use it the first time.